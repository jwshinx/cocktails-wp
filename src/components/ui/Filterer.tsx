import React from 'react'
import { CocktailFilter } from '../../models/cocktail'

interface FilterProps<T> {
  objects: Array<T>
  properties: CocktailFilter<T>[]
  onFilterChange: (property: CocktailFilter<T>) => void
}

function Filterer<T>(props: FilterProps<T>) {
  const { properties, objects, onFilterChange } = props

  if (objects === undefined || objects === null) {
    return <></>
  }
  const object = objects.length > 0 ? objects[0] : {}

  const onCheckboxChange = (filterProperty: CocktailFilter<T>) => {
    // console.log("+++> Filter onCheckboxChange:", filterProperty);
    onFilterChange(filterProperty)
  }

  return (
    <div className="col-12">
      {Object.keys(object).map((key) => {
        return (
          <div key={key} className="row">
            <div
              className="form-check col-6 ml-5 pl-5"
              style={{ paddingLeft: '30px' }}
            >
              <input
                className="form-check-input"
                type="checkbox"
                id={`${key}-true`}
                value={key}
                onChange={() => {
                  onCheckboxChange({
                    property: key as keyof T,
                    isTruthySelected: true,
                  })
                }}
                checked={properties.some(
                  (property: CocktailFilter<T>) =>
                    property.property === key && property.isTruthySelected
                )}
              />
              <label htmlFor={`${key}-true`} className="form-check-label">
                {key} truthy
              </label>
            </div>

            <div className="form-check col-6">
              <input
                style={{ marginLeft: '3px', marginRight: '3px' }}
                className="form-check-input"
                onChange={() => {
                  onCheckboxChange({
                    property: key as keyof T,
                    isTruthySelected: false,
                  })
                }}
                checked={properties.some(
                  (property: CocktailFilter<T>) =>
                    property.property === key && !property.isTruthySelected
                )}
                type="checkbox"
                id={`${key}-false`}
                value={key}
              />
              <label htmlFor={`${key}-false`} className="form-check-label">
                {key} falsy
              </label>
            </div>
            <br />
          </div>
        )
      })}
    </div>
  )
}

export default Filterer
