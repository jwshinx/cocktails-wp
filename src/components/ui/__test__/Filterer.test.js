/**
 * @jest-environment jsdom
 */

import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'

import Filterer from '../Filterer'

const mockFn = jest.fn(() => true)

describe('when GeneralDrink', () => {
  const generalDrinkData = [
    { idDrink: '1', strDrink: 'Martini' },
    { idDrink: '2', strDrink: 'Greyhound' },
    { idDrink: '3', strDrink: 'Mai Tai' },
    { idDrink: '4', strDrink: 'Rob Roy' },
  ]

  const generalDrinkProperties = [
    { property: { idDrink: '1', strDrink: 'Martini' }, isTruthySelected: true },
  ]

  it('should render 4 checkboxes/labels', () => {
    render(
      <Filterer
        objects={generalDrinkData}
        properties={generalDrinkProperties}
        onFilterChange={mockFn}
      />
    )
    const checkboxes = screen.getAllByRole('checkbox')
    expect(checkboxes.length).toBe(4)
    expect(screen.getByLabelText('idDrink truthy')).toBeInTheDocument()
    expect(screen.getByLabelText('idDrink falsy')).toBeInTheDocument()
    expect(screen.getByLabelText('strDrink truthy')).toBeInTheDocument()
    expect(screen.getByLabelText('strDrink falsy')).toBeInTheDocument()
  })

  it('should snapshot GeneralDrink Filterer', () => {
    const tree = renderer
      .create(
        <Filterer
          objects={generalDrinkData}
          properties={generalDrinkProperties}
          onFilterChange={mockFn}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('when DetailDrink', () => {
  const detailDrinkData = [
    { idDrink: '1', strDrink: 'Martini', strIngredient1: 'gin' },
    { idDrink: '2', strDrink: 'Greyhound', strIngredient1: 'vodka' },
    { idDrink: '3', strDrink: 'Mai Tai', strIngredient1: 'rum' },
    { idDrink: '4', strDrink: 'Rob Roy', strIngredient1: 'whiskey' },
  ]

  const detailDrinkProperties = [
    {
      property: { idDrink: '1', strDrink: 'Martini', strIngredient1: 'gin' },
      isTruthySelected: true,
    },
  ]

  it('should render 6 checkboxes/labels', () => {
    render(
      <Filterer
        objects={detailDrinkData}
        properties={detailDrinkProperties}
        onFilterChange={mockFn}
      />
    )
    const checkboxes = screen.getAllByRole('checkbox')
    expect(checkboxes.length).toBe(6)
    expect(screen.getByLabelText('idDrink truthy')).toBeInTheDocument()
    expect(screen.getByLabelText('idDrink falsy')).toBeInTheDocument()
    expect(screen.getByLabelText('strDrink truthy')).toBeInTheDocument()
    expect(screen.getByLabelText('strDrink falsy')).toBeInTheDocument()
    expect(screen.getByLabelText('strIngredient1 truthy')).toBeInTheDocument()
    expect(screen.getByLabelText('strIngredient1 falsy')).toBeInTheDocument()
  })

  it('should snapshot DetailDrink Filterer', () => {
    const tree = renderer
      .create(
        <Filterer
          objects={detailDrinkData}
          properties={detailDrinkProperties}
          onFilterChange={mockFn}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
