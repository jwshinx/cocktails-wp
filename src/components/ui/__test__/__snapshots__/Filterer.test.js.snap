// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when DetailDrink should snapshot DetailDrink Filterer 1`] = `
<div
  className="col-12"
>
  <div
    className="row"
  >
    <div
      className="form-check col-6 ml-5 pl-5"
      style={
        Object {
          "paddingLeft": "30px",
        }
      }
    >
      <input
        checked={false}
        className="form-check-input"
        id="idDrink-true"
        onChange={[Function]}
        type="checkbox"
        value="idDrink"
      />
      <label
        className="form-check-label"
        htmlFor="idDrink-true"
      >
        idDrink
         truthy
      </label>
    </div>
    <div
      className="form-check col-6"
    >
      <input
        checked={false}
        className="form-check-input"
        id="idDrink-false"
        onChange={[Function]}
        style={
          Object {
            "marginLeft": "3px",
            "marginRight": "3px",
          }
        }
        type="checkbox"
        value="idDrink"
      />
      <label
        className="form-check-label"
        htmlFor="idDrink-false"
      >
        idDrink
         falsy
      </label>
    </div>
    <br />
  </div>
  <div
    className="row"
  >
    <div
      className="form-check col-6 ml-5 pl-5"
      style={
        Object {
          "paddingLeft": "30px",
        }
      }
    >
      <input
        checked={false}
        className="form-check-input"
        id="strDrink-true"
        onChange={[Function]}
        type="checkbox"
        value="strDrink"
      />
      <label
        className="form-check-label"
        htmlFor="strDrink-true"
      >
        strDrink
         truthy
      </label>
    </div>
    <div
      className="form-check col-6"
    >
      <input
        checked={false}
        className="form-check-input"
        id="strDrink-false"
        onChange={[Function]}
        style={
          Object {
            "marginLeft": "3px",
            "marginRight": "3px",
          }
        }
        type="checkbox"
        value="strDrink"
      />
      <label
        className="form-check-label"
        htmlFor="strDrink-false"
      >
        strDrink
         falsy
      </label>
    </div>
    <br />
  </div>
  <div
    className="row"
  >
    <div
      className="form-check col-6 ml-5 pl-5"
      style={
        Object {
          "paddingLeft": "30px",
        }
      }
    >
      <input
        checked={false}
        className="form-check-input"
        id="strIngredient1-true"
        onChange={[Function]}
        type="checkbox"
        value="strIngredient1"
      />
      <label
        className="form-check-label"
        htmlFor="strIngredient1-true"
      >
        strIngredient1
         truthy
      </label>
    </div>
    <div
      className="form-check col-6"
    >
      <input
        checked={false}
        className="form-check-input"
        id="strIngredient1-false"
        onChange={[Function]}
        style={
          Object {
            "marginLeft": "3px",
            "marginRight": "3px",
          }
        }
        type="checkbox"
        value="strIngredient1"
      />
      <label
        className="form-check-label"
        htmlFor="strIngredient1-false"
      >
        strIngredient1
         falsy
      </label>
    </div>
    <br />
  </div>
</div>
`;

exports[`when GeneralDrink should snapshot GeneralDrink Filterer 1`] = `
<div
  className="col-12"
>
  <div
    className="row"
  >
    <div
      className="form-check col-6 ml-5 pl-5"
      style={
        Object {
          "paddingLeft": "30px",
        }
      }
    >
      <input
        checked={false}
        className="form-check-input"
        id="idDrink-true"
        onChange={[Function]}
        type="checkbox"
        value="idDrink"
      />
      <label
        className="form-check-label"
        htmlFor="idDrink-true"
      >
        idDrink
         truthy
      </label>
    </div>
    <div
      className="form-check col-6"
    >
      <input
        checked={false}
        className="form-check-input"
        id="idDrink-false"
        onChange={[Function]}
        style={
          Object {
            "marginLeft": "3px",
            "marginRight": "3px",
          }
        }
        type="checkbox"
        value="idDrink"
      />
      <label
        className="form-check-label"
        htmlFor="idDrink-false"
      >
        idDrink
         falsy
      </label>
    </div>
    <br />
  </div>
  <div
    className="row"
  >
    <div
      className="form-check col-6 ml-5 pl-5"
      style={
        Object {
          "paddingLeft": "30px",
        }
      }
    >
      <input
        checked={false}
        className="form-check-input"
        id="strDrink-true"
        onChange={[Function]}
        type="checkbox"
        value="strDrink"
      />
      <label
        className="form-check-label"
        htmlFor="strDrink-true"
      >
        strDrink
         truthy
      </label>
    </div>
    <div
      className="form-check col-6"
    >
      <input
        checked={false}
        className="form-check-input"
        id="strDrink-false"
        onChange={[Function]}
        style={
          Object {
            "marginLeft": "3px",
            "marginRight": "3px",
          }
        }
        type="checkbox"
        value="strDrink"
      />
      <label
        className="form-check-label"
        htmlFor="strDrink-false"
      >
        strDrink
         falsy
      </label>
    </div>
    <br />
  </div>
</div>
`;
